pygmt/src/_common.py

```diff
@@ -39,3 +41,210 @@ def _data_geometry_is_point(data: Any, kind: str) -> bool:
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Invalid component 'invalid' for ... 'mt'.
+
+    >>> _FocalMechanismConvention.from_params(["strike", "dip", "rake"])
+    Traceback (most recent call last):
+        ...
+    pygmt.exceptions.GMTInvalidInput: Fail to determine ...
+    """
+
+    # Mapping of focal mechanism conventions to their single-letter codes.
+    _conventions: ClassVar = {
+        "aki": "a",
+        "gcmt": "c",
+        "partial": "p",
+        "mt": {"full": "m", "deviatoric": "z", "dc": "d"},
+        "principal_axis": {"full": "x", "deviatoric": "t", "dc": "y"},
+    }
+
+    # Mapping of single-letter codes to focal mechanism convention names
+    _codes: ClassVar = {
+        "a": "aki",
+        "c": "gcmt",
+        "p": "partial",
+        "m": "mt",
+        "z": "mt",
+        "d": "mt",
+        "x": "principal_axis",
+        "t": "principal_axis",
+        "y": "principal_axis",
+    }
+
+    # Mapping of focal mechanism conventions to their parameters.
+    _params: ClassVar = {
+        "aki": ["strike", "dip", "rake", "magnitude"],
+        "gcmt": [
+            "strike1",
+            "dip1",
+            "rake1",
+            "strike2",
+            "dip2",
+            "rake2",
+            "mantissa",
+            "exponent",
+        ],
+        "partial": ["strike1", "dip1", "strike2", "fault_type", "magnitude"],
+        "mt": ["mrr", "mtt", "mff", "mrt", "mrf", "mtf", "exponent"],
+        "principal_axis": [
+            "t_value",
+            "t_azimuth",
+            "t_plunge",
+            "n_value",
+            "n_azimuth",
+            "n_plunge",
+            "p_value",
+            "p_azimuth",
+            "p_plunge",
+            "exponent",
+        ],
+    }
+
+    def __init__(
+        self,
+        convention: Literal["aki", "gcmt", "partial", "mt", "principal_axis"],
+        component: Literal["full", "deviatoric", "dc"] = "full",
+    ):
+        """
+        Initialize the FocalMechanismConvention object.
+        """
+        if convention in self._conventions:
+            # Convention is given via 'convention' and 'component' parameters.
+            if component not in {"full", "deviatoric", "dc"}:
+                msg = (
+                    f"Invalid component '{component}' for focal mechanism convention "
+                    f"'{convention}'."
+                )
+                raise GMTInvalidInput(msg)
+
+            self.convention = convention
+            self.code = self._conventions[convention]
+            if isinstance(self.code, dict):
+                self.code = self.code[component]
+        elif convention in self._codes:
+            # Convention is given as a single-letter code.
+            self.code = convention
+            self.convention = self._codes[convention]
+        else:
+            msg = f"Invalid focal mechanism convention '{convention}'."
+            raise GMTInvalidInput(msg)
```

Comment:
Wondering if some of the pattern matching on 'convention' long names (e.g. `akl`) or single-letter aliases (e.g. `a`) could be simplified using [`StrEnum`](https://docs.python.org/3/library/enum.html#enum.StrEnum), but not sure how to handle the extra logic around 'component's. Haven't looked into it too closely, but there's supposedly ways to extend Enums with extra methods - https://realpython.com/python-enum/#adding-and-tweaking-member-methods.

Reply:
Yes, using `StrEnum` can simplify the code logic. The `_FocalMechanismConvention` class has been refactored in https://github.com/GenericMappingTools/pygmt/pull/3551/commits/c59b8a32d9e2219944f6034b73377f59c09307f5.